rs/replicated_state/src/canister_state/system_state/log_memory_store/index_table.rs

```diff
@@ -0,0 +1,568 @@
+    memory::{MemoryPosition, MemorySize},
+    ring_buffer::INDEX_ENTRY_SIZE,
+};
+use crate::page_map::PAGE_SIZE;
+use ic_management_canister_types_private::FetchCanisterLogsFilter;
+
+/// Sentinel value for invalid index entries.
+const INVALID_INDEX_ENTRY: u64 = u64::MAX;
+
+/// Lightweight descriptor of a single log record used by the index table.
+/// It stores the record’s start position in the data region and basic
+/// metadata for range queries (idx, timestamp, byte length).
+#[derive(Debug, Clone, Copy, PartialEq)]
+pub struct IndexEntry {
+    /// Start position of the record within the data region.
+    pub position: MemoryPosition,
+    /// Record unituque sequential index.
+    pub idx: u64,
+    /// Record timestamp.
+    pub timestamp: u64,
+    /// Record total byte length.
+    pub bytes_len: u32,
+}
+const _: () = assert!(INDEX_ENTRY_SIZE.get() as usize == 8 + 8 + 8 + 4);
+
+impl IndexEntry {
+    /// Creates an entry for a record at the given position with its indexing metadata.
+    pub fn new(position: MemoryPosition, record: &LogRecord) -> Self {
+        Self {
+            position,
+            idx: record.idx,
+            timestamp: record.timestamp,
+            bytes_len: record.bytes_len() as u32,
+        }
+    }
+
+    /// Creates an explicitly invalid entry, used to initialize empty slots.
+    fn invalid() -> Self {
+        Self {
+            position: MemoryPosition::new(INVALID_INDEX_ENTRY),
+            idx: INVALID_INDEX_ENTRY,
+            timestamp: INVALID_INDEX_ENTRY,
+            bytes_len: 0,
+        }
+    }
+
+    /// True when this entry contains a valid pointer to a record.
+    pub fn is_valid(&self) -> bool {
+        self.idx != INVALID_INDEX_ENTRY
+    }
+}
+
+/// Index table used to speed up searches in the log’s data region.
+///
+/// The data region is a large ring buffer that stores log records. Because only
+/// the head and tail positions are known, scanning the buffer to locate a
+/// specific record would normally require reading many records sequentially.
+///
+/// The index table solves this by dividing the data region into fixed-size
+/// segments. Each segment has an `IndexEntry` pointing to the most recent record
+/// written in that segment. The table also tracks the position of the oldest
+/// live record (`front`).
+///
+/// Together, these entries let the system quickly determine which parts of the
+/// ring buffer may contain records of interest, reducing the amount of data
+/// that must be read during large or targeted searches.
+#[derive(Debug)]
+pub struct IndexTable {
+    front: Option<IndexEntry>,   // Position of the oldest live log record.
+    entries: Vec<IndexEntry>,    // Array of entries covering all data region segments.
+    segment_size: MemorySize,    // Size of each data region segment in bytes.
+    result_max_size: MemorySize, // Maximum size of results to return.
+    data_capacity: MemorySize,   // Total capacity of the data region.
+}
```

Comment:
Did you also consider storing each `IndexEntry` instead of using these segments? It seems like it would simplify things here and I'm not sure if the space optimization is worth it, especially since the canister should be paying for this space anyway, right?